GVFS/GVFS.Common/Database/PlaceholdersTable.cs

```diff
@@ -0,0 +1,202 @@
+                        {
+                            filePlaceholders.Add(data);
+                        }
+                        else
+                        {
+                            folderPlaceholders.Add(data);
+                        }
+                    }
+                }
+            }
+        }
+
+        public HashSet<string> GetAllFilePaths()
+        {
+            using (IPooledConnection pooled = this.connectionPool.GetConnection())
+            using (IDbCommand command = pooled.Connection.CreateCommand())
+            {
+                HashSet<string> fileEntries = new HashSet<string>();
+                command.CommandText = "SELECT path FROM Placeholders WHERE pathType = 0;";
+                using (IDataReader reader = command.ExecuteReader())
+                {
+                    while (reader.Read())
+                    {
+                        fileEntries.Add(reader.GetString(0));
+                    }
+                }
+
+                return fileEntries;
+            }
+        }
+
+        public void AddPlaceholderData(IPlaceholderData data)
+        {
+            if (data.IsFolder)
+            {
+                if (data.IsExpandedFolder)
+                {
+                    this.AddExpandedFolder(data.Path);
+                }
+                else if (data.IsPossibleTombstoneFolder)
+                {
+                    this.AddPossibleTombstoneFolder(data.Path);
+                }
+                else
+                {
+                    this.AddPartialFolder(data.Path);
+                }
+            }
+            else
+            {
+                this.AddFile(data.Path, data.Sha);
+            }
+        }
+
+        public void AddFile(string path, string sha)
+        {
+            using (IPooledConnection pooled = this.connectionPool.GetConnection())
+            using (IDbCommand command = pooled.Connection.CreateCommand())
+            {
+                Insert(command, new PlaceholderData() { Path = path, PathType = PlaceholderData.PlaceholderType.File, Sha = sha });
+            }
+        }
+
+        public void AddPartialFolder(string path)
+        {
+            using (IPooledConnection pooled = this.connectionPool.GetConnection())
+            using (IDbCommand command = pooled.Connection.CreateCommand())
+            {
+                Insert(command, new PlaceholderData() { Path = path, PathType = PlaceholderData.PlaceholderType.PartialFolder });
+            }
+        }
+
+        public void AddExpandedFolder(string path)
+        {
+            using (IPooledConnection pooled = this.connectionPool.GetConnection())
+            using (IDbCommand command = pooled.Connection.CreateCommand())
+            {
+                Insert(command, new PlaceholderData() { Path = path, PathType = PlaceholderData.PlaceholderType.ExpandedFolder });
+            }
+        }
+
+        public void AddPossibleTombstoneFolder(string path)

```

Comment:
With all these repeated `Add[Type]Folder(string path)` methods, with nearly identical bodies, perhaps a single method `private void AddFolder(string path, PlaceholderData.PlaceholderType type)` would be cleaner, and have these `public` methods forward to the `private` one.